site/docs/design-docs/015-mqtt-connector.md

```diff
@@ -0,0 +1,305 @@
+need to use a unique identifier for the client.
+
+The source itself is not fault-tolerant and does not save any state. In
+case of a restart, the source does not know where it left and relies on
+the broker. If the broker keeps a session for the client (above
+situation), the source continues where it left otherwise the source
+emits messages after the subscription.
+
+Paho client has an `autoReconnect` option, in case of a disconnect, the
+client tries to reconnect to the broker. After the reconnection, source
+re-subscribes to the topics.
+
+### Sink
+
+We use `SinkBuilder` to create a sink for MQTT messages. The sink
+creates a client for each processor and publishes messages to the
+specified topic. We append the global processor index to the specified
+clientId for uniqueness, e.g `producer-1`, `producer-2`...
+
+#### API
+
+Since there are several configuration options for the sink, we created
+a `MqttSinkBuilder` to configure and build the sink.
+
+Below is a usage example of the sink builder with all the configuration
+options:
+
+```java
+MqttSinks.builder()
+        .clientId("producer")
+        .broker("tcp://localhost:1883")
+        .auth("username", "password".toCharArray())
+        .topic("topic")
+        .autoReconnect()
+        .keepSession()
+        .retryCount(5)
+        .messageFn(item -> {
+            MqttMessage message = new MqttMessage(item.getBytes());
+            message.setQos(2);
+            return message;
+        })
+        .build();
+```
+
+You can provide a `MqttConnectOptions` function instead of configuring
+the options one by one.
+
+```java
+MqttSinks.builder()
+        .clientId("producer")
+        .broker("tcp://localhost:1883")
+        .connectOptionsFn(() -> {
+            MqttConnectOptions options = new MqttConnectOptions();
+            options.setCleanSession(false);
+            options.setAutomaticReconnect(true);
+            options.setUserName("username");
+            options.setPassword("password".toCharArray());
+            options.setMaxInflight(100);
+            return options;
+        })
+        .retryFn((exception, tryCount) -> {
+            if(tryCount > 5) {
+                return 0;
+            }
+            if(exception.getReasonCode() != MqttException.REASON_CODE_BROKER_UNAVAILABLE) {
+                return 0;
+            }
+            return 200;
+        })
+        .topic("topic")
+        .messageFn(item -> {
+            MqttMessage message = new MqttMessage(item.getBytes());
+            message.setQos(2);
+            return message;
+        })
+        .build();
+```
+
+#### Fault Tolerance
+
+The sink is not fault-tolerant and does not save any state. In case of
+a restart, some messages can be duplicated.
+
+#### Error handling
+
+The sink uses sync client to publish the messages. Any error/exception
+encountered while publishing the messages will fail the job. User can
+configure the retrying of the messages by providing a retry function.
```

Comment:
I don't like this retry function, because it's a new abstraction for something we already have an abstraction for: `com.hazelcast.jet.retry.RetryStrategy`. Why not use that?